### PR TITLE
unifying the crypto keypair APIs

### PR DESCRIPTION
**Pull request by jakubDoka**
Related to https://github.com/ziglang/zig/pull/20953

I changed all keypair APIs I could find to `init` for the pure derivation from seed, and added `initWithRandomSeed` that's mostly used in tests/benchmarks.

Note: I noticed docs on the new functions are missing, though, I feel like they are unnecessary.

### CODE REVIEW

File: lib/std/crypto/25519/ed25519.zig [hunk 1]
```diff
@@ -506,7 +507,7 @@ test "signature" {
 test "batch verification" {
     var i: usize = 0;
     while (i < 100) : (i += 1) {
-        const key_pair = try Ed25519.KeyPair.create(null);
+        const key_pair = try Ed25519.KeyPair.initWithRandomSeed();
```
Comment (andrewrk) [inline lib/std/crypto/25519/ed25519.zig:510]: unit tests should not use `std.crypto.random`, instead they should use `std.testing.random_seed`.

this is a problem with master branch, but it should be fixed in this PR before merging.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks! Let's take this opportunity to also fix something else that I noticed.

File: lib/std/crypto/25519/ed25519.zig (hunk 1 above)
Comment (jakubDoka) [reply lib/std/crypto/25519/ed25519.zig:510]: I noticed https://github.com/ziglang/zig/blob/master/lib/std/crypto/salsa20.zig#L539 will need to be changed to take a seed, the `initWithRandomSeed` should probably be removed to reduce the likelihood of similar mistakes.

**Comment by jakubDoka** [comment]
I have repeated the `initTestKeypair` to keep it private and also moved the `seed_length` into the `KeyPair` scope.

**Comment by jedisct1** [comment]
I strongly disagree with these changes.  

Virtually all applications should always use non-deterministic keys. We should absolutely not encourage people to provide their own seed, possibly from insecure random number generators or constant seeds. This is a recipe for disaster.

Aa `create()` function that doesn't take any seed for the common case, and `createDeterministic()` for deterministic keys would be way better.

File: lib/std/crypto/25519/x25519.zig [hunk 2]
```diff
@@ -19,8 +19,6 @@ pub const X25519 = struct {
     pub const public_length = 32;
     /// Length (in bytes) of the output of the DH function.
     pub const shared_length = 32;
-    /// Seed (for key pair creation) length in bytes.
-    pub const seed_length = 32;
```
Comment (jedisct1) [inline lib/std/crypto/25519/x25519.zig:23]: That shouldn't be removed.

File: lib/std/crypto/ml_kem.zig [hunk 3]
```diff
@@ -229,8 +229,6 @@ fn Kyber(comptime p: Params) type {
         pub const shared_length = common_shared_key_size;
         /// Length (in bytes) of a seed for deterministic encapsulation.
         pub const encaps_seed_length = common_encaps_seed_length;
-        /// Length (in bytes) of a seed for key generation.
-        pub const seed_length: usize = inner_seed_length + shared_length;
```
Comment (jedisct1) [inline lib/std/crypto/ml_kem.zig:233]: Why remove these useful constants?

File: lib/std/crypto/25519/x25519.zig (hunk 2 above)
Comment (jakubDoka) [reply lib/std/crypto/25519/x25519.zig:23]: It's was moved to the keypair to unify things

File: lib/std/crypto/ml_kem.zig (hunk 3 above)
Comment (jakubDoka) [reply lib/std/crypto/ml_kem.zig:233]: Same here

**Comment by andrewrk** [comment]
> Virtually all applications should always use non-deterministic keys.

Testing is an extremely common use case. 

> We should absolutely not encourage people to provide their own seed, possibly from insecure random number generators or constant seeds. This is a recipe for disaster.

You really think people are going to reach for Ed25519 encryption, and then pass the system time or something like that? ...you know what now that I think of it, [yes, people are absolutely that stupid](https://github.com/llvm/llvm-project/blob/llvmorg-18.1.8/compiler-rt/lib/fuzzer/FuzzerDriver.cpp#L795-L812). Point taken.

> Aa `create()` function that doesn't take any seed for the common case, and `createDeterministic()` for deterministic keys would be way better.

`create` typically means "heap allocated" or allocation of another resource which is not the case here. As a rule of thumb, if there is no "destroy" then there should be no "create".

However, let us accomplish your goal while keeping existing conventions, with `init` and `initDeterministic`.

**Comment by jakubDoka** [comment]
Besides humans, AI now also writes code so this is even more important 😂. I'll get to this tomorrow

**Comment by jedisct1** [comment]
> `create` typically means "heap allocated" or allocation of another resource which is not the case here. As a rule of thumb, if there is no "destroy" then there should be no "create".

Not really. Maybe in other languages, but in Zig, `init()` is typically what gets an allocator, with a corresponding `deinit()` function.

`init()` also conveys the idea that the result with be initialized with default values, not with something different every time.

There's nothing wrong with `create()`. This actually creates a new key every time.

File: lib/std/crypto/benchmark.zig [hunk 4]
```diff
@@ -138,9 +138,15 @@ const signatures = [_]Crypto{
...
+fn initBenchKeyPair(comptime Scheme: type) !Scheme.KeyPair {
+    var seed: [Scheme.KeyPair.seed_length]u8 = undefined;
+    prng.fill(&seed);
+    return try Scheme.KeyPair.createDeterministic(seed);
+}
+
 pub fn benchmarkSignature(comptime Signature: anytype, comptime signatures_count: comptime_int) !u64 {
     const msg = [_]u8{0} ** 64;
-    const key_pair = try Signature.KeyPair.create(null);
+    const key_pair = try initBenchKeyPair(Signature);
```
Comment (alterstep) [inline lib/std/crypto/benchmark.zig:149]: The benchmark can simply use `Signature.KeyPair.create()`.

File: lib/std/crypto/25519/ed25519.zig [hunk 5]
```diff
@@ -503,10 +504,17 @@ test "signature" {
...
 }
 
+fn initTestKeypair(comptime Scheme: type, rng: *std.Random.DefaultPrng) !Scheme.KeyPair {
+    var seed: [Scheme.KeyPair.seed_length]u8 = undefined;
+    rng.fill(&seed);
+    return try Scheme.KeyPair.createDeterministic(seed);
+}
+
 test "batch verification" {
+    var prng = std.Random.DefaultPrng.init(std.testing.random_seed);
```
Comment (alterstep) [inline lib/std/crypto/25519/ed25519.zig:514]: People frequently look at tests to understand how to use things. So maybe use `DefaultCsprng` instead, and rename `prng` to `test_prng` to make it clear that this code is only for testing?

**Comment by andrewrk** [comment]
> Not really. Maybe in other languages, but in Zig, `init()` is typically what gets an allocator, with a corresponding `deinit()` function.

Frank, please do a couple greps and see that this is completely wrong.

**Comment by jakubDoka** [comment]
And once again I am renaming things🍷

**Comment by andrewrk** [comment]
@jakubDoka I recommend for you to sit back and do nothing until we work this out.

**Comment by jakubDoka** [comment]
So, does the `create/destroy` pattern signifies that the object holds an allocation for the full lifetime, whereas `init/deinit` hints that the initialization never allocates and/or allocation is not required for the object to exist?

**Comment by andrewrk** [comment]
The `create` convention generally indicates that the object itself is heap allocated and the return value will be a pointer.

`init` does not hint anything other than that the return type will be of the parent namespace type and not be a pointer.

**Comment by jedisct1** [comment]
Would `generate()` work? This is quite a common name for creating keys.

**Comment by jakubDoka** [comment]
`create` -> `generate` & `createDeterministic` -> `derive`

**Comment by andrewrk** [comment]
"generate" is fine. Mainly, I care about separate functions rather than optionals (the pattern in #20953).

I understand the desire to help people write better code (really, I do), but the bottom line is that entropy is an externally-provided dependency, and it needs to be recognized as one in the API. That's just a fact of reality in programming.

I would like to avoid the convention-breaking `create` name if possible. However if you want to have your own convention in std.crypto, I think that's ... acceptable.

**Comment by jedisct1** [comment]
Let's opt for `generate` and `generateDeterministic`, then.

**Comment by jakubDoka** [comment]
Is this all?

**Pull request closed without merging**